# Final Project Notebook

DS 5001 Text as Data

# Metadata

- Full Name: Shawn Ding, Tianyin Mao
- Userid: exf7sx, qhh3bv
- GitHub Repo URL:[Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)

# Overview

The goal of the final project is for you to create a **digital analytical edition** of a corpus using the tools, practices, and perspectives you’ve learning in this course. You will select a corpus that has already been digitized and transcribed, parse that into an F-compliant set of tables, and then generate and visualize the results of a series of fitted models. You will also draw some tentative conclusions regarding the linguistic, cultural, psychological, or historical features represented by your corpus. The point of the exercise is to have you work with a corpus through the entire pipeline from ingestion to interpretation. 

Specifically, you will acquire a collection of long-form texts and perform the following operations:

- **Convert** the collection from their source formats (F0) into a set of tables that conform to the Standard Text Analytic Data Model (F2).
- **Annotate** these tables with statistical and linguistic features using NLP libraries such as NLTK (F3).
- **Produce** a vector representation of the corpus to generate TFIDF values to add to the TOKEN (aka CORPUS) and VOCAB tables (F4).
- **Model** the annotated and vectorized model with tables and features derived from the application of unsupervised methods, including PCA, LDA, and word2vec (F5).
- **Explore** your results using statistical and visual methods.
- **Present** conclusions about patterns observed in the corpus by means of these operations.

When you are finished, you will make the results of your work available in GitHub (for code) and UVA Box (for data). You will submit to Gradescope (via Canvas) a PDF version of a Jupyter notebook that contains the information listed below.

# Some Details

- Please fill out your answers in each task below by editing the markdown cell. 
- Replace text that asks you to insert something with the thing, i.e. replace `(INSERT IMAGE HERE)` with an image element, e.g. `![](image.png)`.
- For URLs, just paste the raw URL directly into the text area. Don't worry about providing link labels using `[label](link)`.
- Please do not alter the structure of the document or cell, i.e. the bulleted lists. 
- You may add explanatory paragraphs below the bulleted lists.
- Please name your tables as they are named in each task below.
- Tasks are indicated by headers with point values in parentheses.

# Raw Data

## Source Description (1)
Provide a brief description of your source material, including its provenance and content. Tell us where you found it and what kind of content it contains.

### Answer
This corpus is based on the English translation of the I Ching (Book of Changes) by James Legge (1899), obtained from https://sacred-texts.com/ich/index.htm. The dataset consists of 64 hexagrams, each representing a symbolic state or philosophical condition. Each document contains interpretive text explaining the meaning, implications, and guidance associated with each hexagram. The corpus provides rich material for analyzing philosophical language, symbolic reasoning, and patterns in classical Chinese thought translated into English.


1. The 64 hexagrams of the I-Ching are traditionally divided into two canons:

- Upper Canon (hexagrams 1–30): Opens with Qian (☰ Heaven) and Kun (☷ Earth), the two primordial forces. This section deals with cosmological principles — the fundamental nature of heaven, earth, and the universe.
- Lower Canon (hexagrams 31–64): Opens with Xian (Influence) and Heng (Perseverance), and shifts toward human affairs — relationships, social dynamics, family, and governance.

2. Each hexagram is formed by stacking two trigrams:
- Upper trigram (outer) represents the external situation — the environment, social forces, and circumstances you are responding to but cannot fully control.
- Lower trigram (inner) represents the internal condition — your own position, disposition, and the foundation from which you act.

![Trigrams](src/trigram-chart-large2.gif)

## Source Features (1)

Add values for the following items. (Do this for all following bulleted lists.)

- Source URL: https://sacred-texts.com/ich/index.htm

- Number of raw documents: 75 (Hexagrams & Appendix)

- Total size of raw documents: 975 KB (based on the provided i_ching_data.csv)

- File format(s): CSV (parsed from HTML)

## Source Document Structure (1)

Provide a brief description of the internal structure of each document. That, describe the typical elements found in document and their relation to each other. For example, a corpus of letters might be described as having a date, an addressee, a salutation, a set of content paragraphs, and closing. If they are various structures, state that.

### Answer
Each document represents a single hexagram/Appendix and consists of a title, a symbolic representation, and multiple paragraphs of explanatory text. The content typically includes philosophical interpretations, moral guidance, and situational advice. While the structure is relatively consistent across documents, the length and depth of interpretation vary between hexagrams.

- Total paragraph rows : 2714 
    - Hexagram paragraphs: 1016 
	- Appendix paragraphs: 1698 
- Total words : 122,708 
- Unique documents : 75 

OHCO structure for CORPUS: doc_id | para_num | sent_num | token_num

# Parsed and Annotated Data

Parse the raw data into the three core tables of your addition: the `LIB`, `CORPUS`, and `VOCAB` tables.

These tables will be stored as CSV files with header rows.

You may consider using `|` as a delimitter.

Provide the following information for each.

## LIB (2)

The source documents the corpus comprises. These may be books, plays, newspaper articles, abstracts, blog posts, etc. 

Note that these are *not* documents in the sense used to describe a bag-of-words representation of a text, e.g. chapter.

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: `|`
- Number of observations: 75
- List of features, including at least three that may be used for model summarization (e.g. date, author, etc.):

['title', 'url', 'hexagram_img', 'doc_type', 'num_paras', 'total_words', 'author', 'year', 'language', 'source_url', 'genre', 'hexagram_id', 'canon', 'modern_name', 'chinese', 'english', 'upper_trigram', 'upper_trigram_symbol', 'upper_trigram_element', 'lower_trigram', 'lower_trigram_symbol', 'lower_trigram_element']

- Average length of each document in characters: 9,176

## CORPUS (2)

The sequence of word tokens in the corpus, indexed by their location in the corpus and document structures.

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: `|`
- Number of observations Between (should be >= 500,000 and <= 2,000,000 observations.): 123,763
- OHCO Structure (as delimitted column names): ['doc_id','para_num','sent_num','token_num']
- Columns (as delimitted column names, including `token_str`, `term_str`, `pos`, and `pos_group`): ['token_str', 'term_str', 'pos', 'pos_group']

## VOCAB (2)

The unique word types (terms) in the corpus.

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter:`|`
- Number of observations: 6,512
- Columns (as delimitted names, including `n`, `p`', `i`, `dfidf`, `porter_stem`, `max_pos` and `max_pos_group`, `stop`): ['n', 'p', 'i', 'stop', 'porter_stem', 'max_pos', 'max_pos_group', 'dfidf']
- Note: Your VOCAB may contain ngrams. If so, add a feature for `ngram_length`.
- List the top 20 significant words in the corpus by DFIDF.
Top 20 Significant Words by DFIDF:
['line', 'subject', 'shows', 'good', 'place', 'great', 'undivided', 'paragraph', 'fortune', 'divided', 'strong', 'lines', 'weak', 'man', 'position', 'correct', 'proper', 'firm', 'hexagram', 'heaven']

# Derived Tables

## BOW (3)

A bag-of-words representation of the CORPUS.

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: `|`
- Bag (expressed in terms of OHCO levels): doc_id
- Number of observations: 33,362
- Columns (as delimitted names, including `n`, `tfidf`): ['doc_id', 'term_str','n']

## DTM (3)

A represenation of the BOW as a sparse count matrix.

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: '|'  
- Bag (expressed in terms of OHCO levels): doc_id

## TFIDF (3)

A Document-Term matrix with TFIDF values.

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: '|'
- Description of TFIDIF formula ($\LaTeX$ OK): 
$$\text{TF-IDF}_{i,j} = \text{TF}_{i,j} \times \text{IDF}_j$$

## Reduced and Normalized TFIDF_L2 (3)

A Document-Term matrix with L2 normalized TFIDF values.

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: '|'
- Number of features (i.e. significant words): 500
- Principle of significant word selection: 

We retain the top 500 most significant terms based on DF‑IDF scores. Given that the dataset contains only 75 documents, maintaining a lower feature-to-document ratio helps improve model quality. Using too many features—such as 1,000 terms—would likely introduce noise rather than additional meaningful information, which can negatively affect dimensionality‑reduction methods like PCA and topic models such as LDA.

# Models

## PCA Components (4)

- UVA Box URL:
- UVA Box URL of the source TFIDF_L2 table:
- GitHub URL for notebook used to create:
- Delimitter: '|'
- Number of components: 10
- Library used to generate: sklearn.decomposition.PCA
- Top 5 positive terms for first component:

```
--- PCA Metadata ---
Top 5 Positive Terms for PC0:
term_str
chapter      0.379077
earth        0.320396
heaven       0.267664
paragraph    0.205298
khwăn        0.188015
```
- Top 5 negative terms for second component:

```
Top 5 Negative Terms for PC1:
term_str
wife       -0.126031
humility   -0.116088
advance    -0.115373
tâ         -0.094445
husband    -0.080829
Name: PC1, dtype: float64
```

## PCA DCM (4)

The document-component matrix generated.

- GitHub URL for notebook used to create:[Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: `|`

In [4]:
import pandas as pd

DCM = pd.read_csv('output/DCM.csv', sep = '|')
DCM

,doc_id,PC0,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9
0,0,0.295867,-0.071347,-0.150134,-0.251560,0.284512,0.223250,-0.241613,0.034502,0.063758,-0.066292
1,1,0.286675,-0.094470,-0.177660,-0.204630,0.071192,-0.141223,0.167407,0.090141,0.144375,-0.046357
2,2,-0.115286,0.053100,-0.221146,0.213703,0.102185,0.034565,0.052835,0.332643,0.005830,-0.125470
3,3,-0.139340,-0.096573,0.001679,-0.138181,0.058308,-0.071249,0.005262,0.060461,0.150569,-0.263944
4,4,-0.128605,-0.064185,0.157487,-0.124788,0.008288,-0.216246,0.005231,0.312864,-0.218691,-0.098041
...,...,...,...,...,...,...,...,...,...,...,...
70,70,0.515103,0.083462,-0.127337,-0.115286,0.143377,0.004683,-0.120947,0.036970,0.110826,-0.016672
71,71,0.470534,0.015491,-0.199773,-0.197388,-0.082858,-0.091105,0.003015,0.072662,0.166769,0.018430
72,72,0.582515,0.012958,-0.039513,-0.068798,-0.108896,0.025411,-0.127829,-0.029029,-0.107389,0.134412
73,73,-0.000931,-0.040820,-0.079388,-0.098073,-0.224641,0.280555,0.202499,0.080849,-0.194512,0.178852


## PCA Loadings (4)

The component-term matrix generated.

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: '|'

In [29]:
LOADINGS = pd.read_csv('output/LOADINGS.csv',sep='|', index_col=0)
LOADINGS

,PC0,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9
term_str,,,,,,,,,,
abide,-0.016424,-0.007326,0.033589,0.026333,-0.021859,-0.002992,-0.014187,0.067299,0.022023,0.015814
able,0.007950,0.000075,0.011563,0.020902,0.013765,0.007120,0.005626,0.020192,0.010174,-0.034118
accordance,0.085296,0.043391,0.107008,0.128609,-0.004814,-0.028282,0.139476,-0.019805,0.080459,-0.072860
according,0.048353,0.028582,-0.005854,0.065012,0.036715,0.004079,0.025966,0.021628,-0.001550,-0.028906
accordingly,0.019033,0.001463,0.041284,-0.003339,0.012789,0.028633,0.024783,-0.033807,-0.002530,-0.043135
...,...,...,...,...,...,...,...,...,...,...
young,0.001340,-0.039401,-0.102686,0.137306,-0.074815,0.008991,-0.051945,0.039886,-0.039441,-0.009478
younger,-0.004213,-0.024347,-0.012354,0.084624,-0.046626,0.164616,-0.004099,0.086762,0.010241,-0.128865
yâo,0.071137,0.019185,-0.017168,-0.011928,0.019969,-0.004963,-0.018255,0.005618,0.013353,-0.002123


## PCA Visualization 1 (4)

Include a scatterplot of documents in the space created by the first two components.

Color the points based on a metadata feature associated with the documents.

Also include a scatterplot of the loadings for the same two components. (This does not need a feature mapped onto color.)

![pca_viz1](output/pca_viz_1.png)


Briefly describe the nature of the polarity you see in the first component:

> PC0 captures the distinction between the appendix commentaries and the main hexagram texts. The positive pole is driven by scholarly and meta-textual vocabulary ("chapter", "paragraph", "sages") characteristic of the Ten Wings' Confucian commentary style, while the negative pole represents the more oracular and symbolic language of the 64 hexagrams themselves

## PCA Visualization 2 (4)

Include a scatterplot of documents in the space created by the second two components.

Color the points based on a metadata feature associated with the documents.

Also include a scatterplot of the loadings for the same two components. (This does not need a feature mapped onto color.)

![pca2](output/pca_viz2.png)


Briefly describe the nature of the polarity you see in the second component:

> PC2 separates two registers of I-Ching vocabulary: the negative pole is defined by social and relational terms — 'wife', 'family', 'union', 'restraint', 'advance' — representing the text's moral philosophy of human relationships and duty. The positive pole is anchored by elemental and passive terms — 'water', 'change', 'pleasure' — associated with yielding, natural process, and the Kan trigram (☵ Water). 

> PC2 thus reflects a fundamental tension within the I-Ching between active moral engagement in the social world and receptive attunement to natural forces. 

> PC3 further refines this by separating active roles ('advance', 'wife' as agent) from passive states ('couch', 'pleasure'), though this axis explains less variance and is harder to interpret definitively

## LDA TOPIC (4)

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: '|'
- Libary used to compute: sklearn.decomposition.LatentDirichletAllocation
- A description of any filtering, e.g. POS (Nouns and Verbs only):

  Terms were filtered in three stages: (1) sklearn's built-in English stop word list was applied to the DTM; (2) corpus-specific structural terms were removed (line, subject, shows, divided, undivided, place, hexagram, etc.) as they appear in virtually every hexagram and carry no topical information; (3) only Nouns (NN) and Verbs (VB) by modal POS tag were retained, with auxiliary verbs (be, is, are, have, has) additionally excluded.

- Number of components: 5
- Any other parameters used: random_state=42 (for reproducibility), max_iter=20, learning_method='online'
- Top 5 words and best-guess labels for topic five topics by mean document weight:

  T00: ['followed', 'things', 'denotes', 'state', 'tâ']
  T01: ['waiting', 'stream', 'success', 'end', 'yin']
  T02: ['fortune', 'correlate', 'firm', 'king', 'strength']
  T03: ['advance', 'fortune', 'correlate', 'contention', 'firm']
  T04: ['man', 'heaven', 'fortune', 'earth', 'things']


In [12]:
from sklearn.decomposition import LatentDirichletAllocation
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

In [21]:
DTM = pd.read_csv('output/DTM.csv', sep='|', index_col=0)
VOCAB = pd.read_csv('output/VOCAB.csv', sep='|', index_col=0)

DTM_LDA = DTM.loc[:, ~DTM.columns.isin(ENGLISH_STOP_WORDS)]

# Remove very rare words (optional but recommended)
DTM_LDA = DTM_LDA.loc[:, DTM_LDA.sum() > 5]

DTM_LDA.head()

,1,10,11,12,13,14,15,16,17,18,...,yî,yû,yü,yüan,zăn,î,žhui,žin,žing,žî
doc_id,,,,,,,,,,,,,,,,,,,,,
0,2,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,2,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,2,0,0,0,0,0,0,0,0,0,...,1,0,0,0,0,0,0,0,0,0
3,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,2,0,0,0,0,0,0,0,0,0,...,1,0,0,0,0,0,0,0,0,0


In [22]:
# Corpus-specific structural words that appear in every hexagram
# and carry no topical information
CORPUS_STOPWORDS = {
    'line', 'lines', 'subject', 'shows', 'show', 'divided', 'undivided',
    'place', 'figure', 'hexagram', 'trigram', 'paragraph', 'appendix',
    'ching', 'said', 'says', 'mean', 'means', 'meaning', 'refer', 'refers',
    'nine', 'six', 'first', 'second', 'third', 'fourth', 'fifth', 'sixth',
    'topmost', 'lowest', 'upper', 'lower', 'inner', 'outer'
}

# Build filtered DTM for LDA:
#   1. Start from DTM_LDA (sklearn stop words already removed)
#   2. Remove corpus-specific stopwords
#   3. Keep only Nouns and Verbs (excluding 'be', 'have' auxiliaries)
AUX_VERBS = {'be','is','are','was','were','been','being',
              'have','has','had','do','does','did'}

v_filter = VOCAB[
    (VOCAB['max_pos_group'].isin(['NN', 'VB'])) &
    (VOCAB['stop'] == 0) &
    (~VOCAB.index.isin(CORPUS_STOPWORDS)) &
    (~VOCAB.index.isin(AUX_VERBS))
].index

# Use DTM_LDA (stop-word filtered) as base, then apply POS filter
DTM_LDA_filtered = DTM_LDA[DTM_LDA.columns.intersection(v_filter)]

# Remove terms appearing in fewer than 2 documents (too rare for LDA)
DTM_LDA_filtered = DTM_LDA_filtered.loc[:, (DTM_LDA_filtered > 0).sum() >= 2]

print(f"Terms for LDA: {DTM_LDA_filtered.shape[1]}")

Terms for LDA: 1244


In [23]:
# Fit LDA
n_topics = 5
lda = LatentDirichletAllocation(
    n_components=n_topics,
    random_state=42,
    max_iter=20,
    learning_method='online'
)

## LDA THETA (4)

- GitHub URL for notebook used to create:[Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: '|'

In [24]:
THETA = pd.DataFrame(
    lda.fit_transform(DTM_LDA_filtered),
    index=DTM_LDA_filtered.index,
    columns=[f"T{str(i).zfill(2)}" for i in range(n_topics)]
)


In [25]:
THETA = pd.DataFrame(
    lda.fit_transform(DTM_LDA_filtered),
    index=DTM_LDA_filtered.index,
    columns=[f"T{str(i).zfill(2)}" for i in range(n_topics)]
)

## LDA PHI (4)

- GitHub URL for notebook used to create:[Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: '|'

In [26]:
PHI = pd.DataFrame(
    lda.components_,
    columns=DTM_LDA_filtered.columns,
    index=[f"T{str(i).zfill(2)}" for i in range(n_topics)]
)

In [27]:
# Report topics
print("\n=== TOP 5 WORDS PER TOPIC ===")
for t in PHI.index:
    words = PHI.loc[t].sort_values(ascending=False).head(5).index.tolist()
    print(f"{t}: {words}")

print("\n=== MEAN DOCUMENT WEIGHT ===")
print(THETA.mean().sort_values(ascending=False))


=== TOP 5 WORDS PER TOPIC ===
T00: ['followed', 'things', 'denotes', 'state', 'tâ']
T01: ['waiting', 'stream', 'success', 'end', 'yin']
T02: ['fortune', 'correlate', 'firm', 'king', 'strength']
T03: ['advance', 'fortune', 'correlate', 'contention', 'firm']
T04: ['man', 'heaven', 'fortune', 'earth', 'things']

=== MEAN DOCUMENT WEIGHT ===
T04    0.570150
T02    0.329069
T00    0.040999
T03    0.035498
T01    0.024284
dtype: float64


## LDA + PCA Visualization (4)

Apply PCA to the THETA table and plot the topics in the space opened by the first two components.

Size the points based on the mean document weight of each topic (using the THETA table).

Color the points basd on a metadata feature from the LIB table.

Provide a brief interpretation of what you see.

![image](output/lda_pca_viz.png)

The LDA+PCA plot reveals that topic structure does not map cleanly onto the Upper/Lower Canon division 
- Upper and Lower Canon documents are largely intermixed across the space, suggesting that thematic content cuts across the traditional structural boundary. The clearest separation is between the appendix texts (green), which cluster distinctly from the hexagrams 

- particularly the isolated outlier at the top, which is dominated by a unique topic corresponding to the scholarly commentary register of the Ten Wings. Point sizes indicate that many hexagrams are strongly dominated by a single topic, while others 

- mostly in the Lower Canon — show more diffuse topic distributions, reflecting their greater thematic complexity.

## Sentiment VOCAB_SENT (4)

Sentiment values associated with a subset of the VOCAB from a curated sentiment lexicon.

- GitHub URL for notebook used to create: [Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: `|`

In [30]:
VOCAB_SENT = pd.read_csv('output/BOW_SENT.csv', sep='|', index_col=0)
VOCAB_SENT

,doc_id,term_str,n,sentiment,weighted_sentiment
0,0,1,2,NaN,NaN
1,0,2,2,NaN,NaN
2,0,3,3,NaN,NaN
3,0,4,1,NaN,NaN
4,0,5,1,NaN,NaN
...,...,...,...,...,...
33357,74,žhui,1,0.0,0.0
33358,74,žieh,1,0.0,0.0
33359,74,žin,1,0.0,0.0
33360,74,žing,1,0.0,0.0


## Sentiment DOC_SENT (4)

Computed sentiment per bag computed from BOW_SENT.

- GitHub URL for notebook used to create:[Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: `|`
- Document bag expressed in terms of OHCO levels: 'doc_id'

![image](output/Sentiment%20Score%20by%20Hexagram.png)

## VOCAB_W2V (4)

A table of word2vec features associated with terms in the VOCAB table.

- GitHub URL for notebook used to create:[Project_Notebook_I_CHING.ipynb](https://github.com/Poptartmeow/DS5001-I-Ching-the-Book-of-Change/blob/3e638782f2bceb6d1caac5a1f8e9a809ebc77cd6/Project_Notebook_I_CHING.ipynb)
- Delimitter: `|`
- Document bag expressed in terms of OHCO levels: doc_id | para_num | sent_num
- Number of features generated: 100
- The library used to generate the embeddings: gensim.models.Word2Vec

In [28]:
VOCAB_W2V = pd.read_csv('output/VOCAB_W2V.csv',sep='|',index_col=0)

print(f"Terms with embeddings : {len(VOCAB_W2V):,}")
print(f"Terms without (excluded): {len(VOCAB_W2V) - len(VOCAB_W2V):,}")
print(f"Vector dimensions     : {VOCAB_W2V.shape[1]}")

Terms with embeddings : 3,957
Terms without (excluded): 0
Vector dimensions     : 100


## Word2vec tSNE Plot (4)

Plot word embedding featues in two-dimensions using t-SNE.

Describe a cluster in the plot that captures your attention.

The t-SNE plot reveals that Word2Vec has recovered three distinct registers within the I-Ching's vocabulary. 
- The moral evaluation cluster (top-left) groups judgment words — 'correct', 'advantageous', 'repentance' — that form the text's guidance apparatus. 
- The structural annotation cluster (bottom-centre) isolates the mechanical line-commentary boilerplate ('second', 'divided', 'shows'), confirming these terms are semantically interchangeable in context.
- The cosmological cluster (far right) places 'heaven', 'earth', 'dragon', 'khien', and 'trigram' in close proximity to scholarly meta-vocabulary from the appendices, suggesting that Legge's commentary engages most intensively with the text's symbolic cosmology. 
- The tight clustering of 'superior' and 'man' confirms the model has learned the I-Ching's central moral actor as a fixed semantic unit.

![image](output/w2v_tsne.png)

# Riffs

Provde at least three visualizations that combine the preceding model data in interesting ways.

These should provide insight into how features in the LIB table are related. 

The nature of this relationship is left open to you -- it may be correlation, or mutual information, or something less well defined. 

In doing so, consider the following visualization types:

- Hierarchical cluster diagrams
- Heatmaps
- Scatter plots
- KDE plots
- Dispersion plots
- t-SNE plots
- etc.

## Riff 1 (5)

![image](output/Hierarchical%20Clustering%20of%20Hexagrams.png)

The Ward-linkage dendrogram reveals that linguistic similarity among hexagrams does not follow the traditional Upper/Lower Canon boundary — hexagrams cluster by thematic vocabulary rather than canonical position. The tightest pairing in the corpus is hexagrams 63 and 64 (Kî Žî and Wei Žî, After and Before Completion), whose near-zero merge distance reflects their well-known structural complementarity. At the other extreme, Hexagram 1 (Khien, The Creative) and the fellowship-themed cluster (Pî, ThungZăn, Žhui) are the most distinctive texts, joining the main tree only at maximum distance. This suggests that the I-Ching's opening hexagram occupies a genuinely unique linguistic register — consistent with its role as the cosmological foundation of the entire system.

## Riff 2 (5)

![image](output/Topic%20Distribution%20Heatmap%20by%20Section.png)

The heatmap with three-way canon grouping reveals a clear topical stratification. The Appendix is most distinctly marked by T00 (0.18), a topic absent from the hexagram texts, confirming that the Ten Wings commentaries occupy a unique linguistic register. T02 emerges as the Lower Canon's signature topic (0.50), compared to 0.25 in the Upper Canon and near-zero in the Appendix — consistent with the Lower Canon's thematic emphasis on human relationships and social situations. T04 dominates all three groups but is most concentrated in the Appendix (0.81) and Upper Canon (0.66), reflecting the more formulaic, structurally repetitive nature of those texts. Together, the three rows map the I-Ching's internal structure onto distinct topical profiles with notable clarity.

## Riff 3 (5)

![image](output/KDE%20Plot%20of%20Sentiment%20Distribution%20by%20Book.png)

The KDE plot reveals three distinct sentiment profiles across the corpus. The Lower Canon (hex 31–64) clusters tightly around a mildly positive mean with minimal negative tail, suggesting its human-centred hexagrams are expressed in more uniformly measured language. The Upper Canon (hex 1–30) shows greater variance, with a wider spread and a left tail extending into negative sentiment — reflecting its broader emotional range from cosmic grandeur to existential peril. Most strikingly, the Appendix occupies an entirely different region of the sentiment space, shifted far rightward with a peak near 0.06. This confirms that the Ten Wings commentaries carry a distinct tonal register: the elevated, appreciative prose of Confucian scholarship reads as substantially more positive than the oracular language of the hexagram texts themselves.

# Interpretation (4)

Describe something interesting about your corpus that you discovered during the process of completing this assignment.

At a minumum, use 250 words, but you may use more. You may also add images if you'd like.

## Our Interpretation

The most surprising discovery from this project is that the I Ching's traditional Upper Canon / Lower Canon division is only a partial description of its actual linguistic structure. Every method in the pipeline — PCA, LDA, hierarchical clustering, and sentiment analysis — told a consistent but more nuanced story: the real boundary in the text runs not between Hexagrams 1–30 and 31–64, but between the hexagram corpus as a whole and the Ten Wings appendices that Legge appended to it.

This became clearest in the Principal Component Analysis. PC0 did not separate the Upper Canon from the Lower Canon — it separated both canons from the Appendix. The loading plot revealed why: the terms driving positive PC0 values are "chapter," "paragraph," "heaven," "earth," and "khien" — the meta-analytical vocabulary of Legge's own commentary prose, not the spare oracular language of the hexagram lines. The Appendix, in other words, talks *about* the I Ching in a register that is linguistically foreign to the hexagram text itself. The LDA heatmap confirmed this: Topic T04 dominates both canons (0.66 Upper, 0.41 Lower), but the Appendix is almost entirely T04 (0.81), with a modest but unique loading on T00 (0.18) that barely registers in the hexagrams. The Appendix is not a continuation of the Book of Changes — it is a commentary tradition that has been translated in the same breath, and the vector space models can tell them apart.

Within the hexagram corpus itself, the analysis recovered finer structure. PC2 (pca_viz_2) exposed a domestic-relational pole — loaded on "wife," "family," "union," "restraint" — opposite a natural-elemental pole loaded on "water," "change," and "pleasure." This is an authentic tension in the I Ching: some hexagrams govern cosmic or military affairs; others address marriage, household, and personal conduct. Intriguingly, this axis does not align cleanly with Upper vs Lower Canon either, which suggests the traditional two-part division is itself an organisational simplification of a more continuous gradient.

The hierarchical clustering (Ward linkage on TF-IDF vectors) reinforced this point most dramatically. The dendrogram groups hexagrams in ways that frequently cross canon boundaries. The tightest merge in the entire tree pairs Wei Žî (LXIV, "Before Completion") with Kî Žî (LXIII, "After Completion") — the final two hexagrams — which is precisely what a scholar of the text would predict: these two have always been understood as a completion dyad. That the algorithm recovered this pairing from raw token frequencies alone, without any domain knowledge, is a satisfying validation of the method. Equally telling is the opposite end: Hexagram I (Khien, "The Creative") sits on the most isolated branch in the tree, merging last with all others. Khien is the origin hexagram, composed of six unbroken yang lines, and its linguistic signature is so distinctive that no other hexagram closely resembles it.

The sentiment analysis added a final layer. The KDE plot showed the Lower Canon as the most tightly peaked and uniformly mild-positive of the three sections, while the Upper Canon had the widest spread — extending into slightly negative territory — suggesting greater emotional range in the earlier, more cosmologically ambitious hexagrams. The Appendix, despite its scholarly register, was the most positive of all, its distribution shifted furthest rightward. Legge's commentary prose turns out to carry the most optimistic tone in the entire volume.

Taken together, these findings suggest that what appears to be a unified ancient text is, computationally, at least two distinct corpora (oracular lines versus philosophical commentary) layered on top of a continuous thematic space that the Upper/Lower division only roughly approximates. Modern text-analytic methods do not merely confirm what scholars already knew — they reveal the grain of the text at a resolution traditional reading cannot easily reach.